src/Microsoft.AspNetCore.Http.Abstractions/Routing/RouteValueDictionary.cs

```diff
@@ -0,0 +1,614 @@
+            }
+
+            public KeyValuePair<string, object> Current { get; private set; }
+
+            object IEnumerator.Current => Current;
+
+            public void Dispose()
+            {
+            }
+
+            public bool MoveNext()
+            {
+                if (++_index < _dictionary.Count)
+                {
+                    if (_dictionary._propertyStorage != null)
+                    {
+                        var storage = _dictionary._propertyStorage;
+                        var property = storage.Properties[_index];
+                        Current = new KeyValuePair<string, object>(property.Name, property.GetValue(storage.Value));
+                        return true;
+                    }
+
+                    Current = _dictionary._arrayStorage[_index];
+                    return true;
+                }
+
+                Current = default;
+                return false;
+            }
+
+            public void Reset()
+            {
+                Current = default;
+                _index = -1;
+            }
+        }
+
+        internal class PropertyStorage
+        {
+            private static readonly PropertyCache _propertyCache = new PropertyCache();
+
+            public readonly object Value;
+            public readonly PropertyHelper[] Properties;
+
+            public PropertyStorage(object value)
+            {
+                Debug.Assert(value != null);
+                Value = value;
+
+                // Cache the properties so we can know if we've already validated them for duplicates.
+                var type = Value.GetType();
+                if (!_propertyCache.TryGetValue(type, out Properties))
+                {
+                    Properties = PropertyHelper.GetVisibleProperties(type);
+                    ValidatePropertyNames(type, Properties);
+                    _propertyCache.TryAdd(type, Properties);
+                }
+            }
+
+            private static void ValidatePropertyNames(Type type, PropertyHelper[] properties)
+            {
+                var names = new Dictionary<string, PropertyHelper>(StringComparer.OrdinalIgnoreCase);
+                for (var i = 0; i < properties.Length; i++)
+                {
+                    var property = properties[i];
+
+                    if (names.TryGetValue(property.Name, out var duplicate))
+                    {
+                        var message = Resources.FormatRouteValueDictionary_DuplicatePropertyName(
+                            type.FullName,
+                            property.Name,
+                            duplicate.Name,
+                            nameof(RouteValueDictionary));
+                        throw new InvalidOperationException(message);
+                    }
+
+                    names.Add(property.Name, property);
+                }
+            }
+        }
+
+        private class PropertyCache : ConcurrentDictionary<Type, PropertyHelper[]>
```

Comment:
This is weird

Reply:
https://github.com/aspnet/Routing/pull/708